File: bin/aclab_skills.py
#!/usr/bin/env python3
"""aclab-skills — cross-platform skill installer (Python stdlib only).

  uvx --from git+https://github.com/austinyuch/skills aclab-skills <agent>
  pipx run --spec git+https://github.com/austinyuch/skills aclab-skills <agent>
  python3 bin/aclab_skills.py <agent>            (from a checkout)

<agent> = opencode | claude | codex | kiro   (default: opencode)
Override the destination with --target <dir> or SKILLS_TARGET=<dir>.

When run as an installed tool (uvx/pipx) outside a checkout, the skill data is
not bundled in the wheel, so the CLI shallow-clones the repo into a cache dir.
"""
from __future__ import annotations

import argparse
import json
import os
import shutil
import subprocess
import sys
from pathlib import Path

REPO_URL = "https://github.com/austinyuch/skills"

# code-review's native CLI lives in GitHub Releases (the repo is private, so the
# download is authenticated via `gh`). Bump CLI_TAG when a new binary is released.
CLI_REPO = "austinyuch/skills"
CLI_TAG = "review-cli-v0.16.1"


def review_cli_asset() -> str | None:
    import platform
    o = {"darwin": "darwin", "linux": "linux", "windows": "windows"}.get(platform.system().lower())
    m = platform.machine().lower()
    a = "arm64" if m in ("arm64", "aarch64") else ("amd64" if m in ("x86_64", "amd64") else None)
    if not o or not a:
        return None
    return f"review-cli-{o}-{a}" + (".exe" if o == "windows" else "")


def fetch_review_cli(target: Path) -> None:
    asset = review_cli_asset()
    dest = target / "code-review" / "scripts"
    if not asset:
        print("   ⚠️  unsupported platform for review-cli"); return
    if not dest.exists():
        print("   ⚠️  code-review not installed — skipping --with-cli"); return
    cmd = ["gh", "release", "download", CLI_TAG, "-R", CLI_REPO, "-p", asset, "-D", str(dest), "--clobber"]
    try:
        print(f"⬇️  fetching {asset} from {CLI_REPO}@{CLI_TAG} (gh) …")
        subprocess.run(cmd, check=True)
        f = dest / asset
        if os.name != "nt":
            f.chmod(0o755)
        print(f"   ✅ review-cli installed: {f}")
    except FileNotFoundError:
        print("   ⚠️  GitHub CLI (gh) not found. Install gh + auth, then:")
        print(f'      gh release download {CLI_TAG} -R {CLI_REPO} -p {asset} -D "{dest}" --clobber')
    except subprocess.CalledProcessError as e:
        print(f"   ⚠️  could not fetch review-cli (gh exit {e.returncode}).")
        print(f'      gh release download {CLI_TAG} -R {CLI_REPO} -p {asset} -D "{dest}" --clobber')

AGENT_HOMES = {
    "opencode": lambda: (Path(os.environ["XDG_CONFIG_HOME"]) / "opencode" / "skills")
    if os.environ.get("XDG_CONFIG_HOME")
    else Path.home() / ".config" / "opencode" / "skills",
    "claude": lambda: Path.home() / ".claude" / "skills",
    "codex": lambda: Path.home() / ".codex" / "skills",
    "kiro": lambda: Path.home() / ".kiro" / "skills",
}


def find_data_root(start: Path) -> Path | None:
    """Walk up from `start` to find a dir holding skills-manifest.json + skills/."""
    for d in [start, *start.parents]:
        if (d / "skills-manifest.json").is_file() and (d / "skills").is_dir():
            return d
    return None


def cache_clone() -> Path:
    """Shallow-clone (or refresh) the repo into a stable cache dir, return its path."""
    base = Path(os.environ.get("XDG_CACHE_HOME", Path.home() / ".cache")) / "aclab-skills"
    repo = base / "repo"
    base.mkdir(parents=True, exist_ok=True)
    if (repo / "skills-manifest.json").is_file():
        subprocess.run(["git", "-C", str(repo), "pull", "--ff-only", "--depth", "1"],
                       check=False, capture_output=True)
    else:
        print(f"⬇️  fetching skills from {REPO_URL} …")
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, str(repo)], check=True)
    return repo


def main() -> int:
    p = argparse.ArgumentParser(prog="aclab-skills", description="Install aclab skills into a coding agent's skill home.")
    p.add_argument("agent", nargs="?", default="opencode", help="opencode | claude | codex | kiro")
    p.add_argument("--target", help="explicit destination dir (wins over <agent>)")
    p.add_argument("--dry-run", action="store_true")
    p.add_argument("--with-cli", action="store_true", help="also fetch code-review's review-cli binary via gh (needs auth)")
    # --layout kept for backward compatibility: flat is the only supported layout
    # (one-level loaders like Claude Code cannot discover a nested category tree).
    p.add_argument("--layout", choices=("flat",), default="flat", help=argparse.SUPPRESS)
    args = p.parse_args()

    target = args.target or os.environ.get("SKILLS_TARGET")
    if not target:
        if args.agent not in AGENT_HOMES:
            p.error(f"unknown agent: {args.agent} (use opencode|claude|codex|kiro, or --target)")
        target = str(AGENT_HOMES[args.agent]())
    target = Path(target)

    root = find_data_root(Path(__file__).resolve().parent)
    if root is None:
        root = cache_clone()
    manifest = json.loads((root / "skills-manifest.json").read_text(encoding="utf-8"))
    source = root / "skills"

    print(f"📦 aclab skills from: {root}")
    print(f"🤖 Agent: {'custom' if args.target or os.environ.get('SKILLS_TARGET') else args.agent}")
    print(f"🎯 Target: {target}{'  (dry-run)' if args.dry_run else ''}\n")
    print("🧭 Layout: flat (single-level)\n")
    if not args.dry_run:
        target.mkdir(parents=True, exist_ok=True)

    installed = missing = 0

    def copy_dir(src: Path, dst: Path, name: str) -> None:
        nonlocal installed, missing
        if not src.is_dir():
            print(f"   ⚠️  missing: {name}"); missing += 1; return
        if not args.dry_run:
            dst.parent.mkdir(parents=True, exist_ok=True)
            if dst.exists():
                shutil.rmtree(dst)
            shutil.copytree(src, dst)
        print(f"   ✅ {name}"); installed += 1

    for skill in manifest.get("skills", []):
        copy_dir(source / skill, target / skill, skill)

    for row in manifest.get("standalone_files", []):
        src = source / row["file"]
        dst = target / row["target_path"]
        if not src.is_file():
            print(f"   ⚠️  missing file: {row['file']}"); missing += 1; continue
        if not args.dry_run:
            dst.parent.mkdir(parents=True, exist_ok=True)
            shutil.copy2(src, dst)
        print(f"   ✅ {dst.relative_to(target)}"); installed += 1

    print("\n" + "━" * 32)
    print(f"📊 {'Would install' if args.dry_run else 'Installed'}: {installed}   ⚠️  Missing: {missing}")
    print("━" * 32)
    print(f"Skills {'would be' if args.dry_run else 'are now'} in: {target}")
    if (target / "code-review").exists() and not args.dry_run:
        if args.with_cli:
            fetch_review_cli(target)
        else:
            print('ℹ️  code-review\'s review-cli binary is not bundled — re-run with --with-cli to fetch it '
                  '(needs gh auth; the repo is private). See README "Native binaries".')
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: bin/aclab-skills.mjs
#!/usr/bin/env node
// aclab-skills — cross-platform skill installer (Node, zero deps).
//
//   npx  -y github:austinyuch/skills <agent>
//   bunx    github:austinyuch/skills <agent>
//   node bin/aclab-skills.mjs <agent>            (from a checkout)
//
// <agent> = opencode | claude | codex | kiro   (default: opencode)
// Override the destination with --target <dir> or SKILLS_TARGET=<dir>.
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { execFileSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// code-review's native CLI lives in GitHub Releases (the repo is private, so the
// download is authenticated via `gh`). Bump CLI_TAG when a new binary is released.
const CLI_REPO = 'austinyuch/skills';
const CLI_TAG = 'review-cli-v0.15.0';

function reviewCliAsset() {
  const o = { darwin: 'darwin', linux: 'linux', win32: 'windows' }[process.platform];
  const a = { x64: 'amd64', arm64: 'arm64' }[process.arch];
  if (!o || !a) return null;
  return `review-cli-${o}-${a}${o === 'windows' ? '.exe' : ''}`;
}

function fetchReviewCli(target) {
  const asset = reviewCliAsset();
  const dest = path.join(target, 'code-review', 'scripts');
  if (!asset) { console.log(`   ⚠️  unsupported platform for review-cli (${process.platform}/${process.arch})`); return; }
  if (!fs.existsSync(dest)) { console.log('   ⚠️  code-review not installed — skipping --with-cli'); return; }
  try {
    console.log(`⬇️  fetching ${asset} from ${CLI_REPO}@${CLI_TAG} (gh) …`);
    execFileSync('gh', ['release', 'download', CLI_TAG, '-R', CLI_REPO, '-p', asset, '-D', dest, '--clobber'], { stdio: 'inherit' });
    if (process.platform !== 'win32') fs.chmodSync(path.join(dest, asset), 0o755);
    console.log(`   ✅ review-cli installed: ${path.join(dest, asset)}`);
  } catch (e) {
    const why = e && e.code === 'ENOENT' ? 'GitHub CLI (gh) not found' : (e.message || e);
    console.log(`   ⚠️  could not fetch review-cli (${why}).`);
    console.log(`      Install gh + auth, then: gh release download ${CLI_TAG} -R ${CLI_REPO} -p ${asset} -D "${dest}" --clobber`);
  }
}

const AGENT_HOMES = {
  opencode: () => process.env.XDG_CONFIG_HOME
    ? path.join(process.env.XDG_CONFIG_HOME, 'opencode', 'skills')
    : path.join(os.homedir(), '.config', 'opencode', 'skills'),
  claude: () => path.join(os.homedir(), '.claude', 'skills'),
  codex:  () => path.join(os.homedir(), '.codex', 'skills'),
  kiro:   () => path.join(os.homedir(), '.kiro', 'skills'),
};

function usage(code = 0) {
  console.log(`aclab-skills — install aclab skills into a coding agent's skill home

Usage:
  aclab-skills <agent> [--target <dir>] [--dry-run]
  <agent> = opencode | claude | codex | kiro   (default: opencode)

Examples:
  npx -y github:austinyuch/skills claude
  bunx github:austinyuch/skills codex
  SKILLS_TARGET=/custom/path npx -y github:austinyuch/skills

Env:  SKILLS_TARGET overrides the destination (wins over <agent>).`);
  process.exit(code);
}

// Find the repo data root (holds skills-manifest.json) by walking up from this file.
function findDataRoot(start) {
  let dir = start;
  for (let i = 0; i < 8; i++) {
    if (fs.existsSync(path.join(dir, 'skills-manifest.json')) &&
        fs.existsSync(path.join(dir, 'skills'))) return dir;
    const up = path.dirname(dir);
    if (up === dir) break;
    dir = up;
  }
  return null;
}

function main() {
  const argv = process.argv.slice(2);
  let agent = null, target = process.env.SKILLS_TARGET || null, dryRun = false, withCli = false;
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '-h' || a === '--help') usage(0);
    else if (a === '--dry-run') dryRun = true;
    else if (a === '--with-cli') withCli = true;
    else if (a === '--target') target = argv[++i];
    else if (a.startsWith('--target=')) target = a.slice('--target='.length);
    else if (!a.startsWith('-') && !agent) agent = a;
  }
  agent = agent || 'opencode';
  if (!target && !(agent in AGENT_HOMES)) {
    console.error(`✖ Unknown agent: ${agent} (use opencode|claude|codex|kiro, or --target)`);
    process.exit(2);
  }
  if (!target) target = AGENT_HOMES[agent]();

  const root = findDataRoot(__dirname);
  if (!root) {
    console.error('✖ Could not locate skills-manifest.json / skills/. Run from a checkout or via `npx github:austinyuch/skills`.');
    process.exit(1);
  }
  const manifest = JSON.parse(fs.readFileSync(path.join(root, 'skills-manifest.json'), 'utf8'));
  const source = path.join(root, 'skills');

  console.log(`📦 aclab skills from: ${root}`);
  console.log(`🤖 Agent: ${target === process.env.SKILLS_TARGET ? 'custom' : agent}`);
  console.log(`🎯 Target: ${target}${dryRun ? '  (dry-run)' : ''}\n`);
  if (!dryRun) fs.mkdirSync(target, { recursive: true });

  let installed = 0, missing = 0;
  const copyDir = (src, dst, name) => {
    if (!fs.existsSync(src)) { console.log(`   ⚠️  missing: ${name}`); missing++; return; }
    if (!dryRun) fs.cpSync(src, dst, { recursive: true });
    console.log(`   ✅ ${name}`); installed++;
  };

  for (const group of ['families', 'categories']) {
    for (const [key, val] of Object.entries(manifest[group] || {})) {
      for (const skill of val.skills || []) {
        copyDir(path.join(source, key, skill), path.join(target, skill), skill);
      }
    }
  }
  for (const row of manifest.standalone_files || []) {
    const src = path.join(source, row.category, row.target_path);
    const dst = path.join(target, row.file);
    if (!fs.existsSync(src)) { console.log(`   ⚠️  missing file: ${row.file}`); missing++; continue; }
    if (!dryRun) { fs.mkdirSync(path.dirname(dst), { recursive: true }); fs.copyFileSync(src, dst); }
    console.log(`   ✅ ${row.file}`); installed++;
  }

  console.log(`\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);
  console.log(`📊 ${dryRun ? 'Would install' : 'Installed'}: ${installed}   ⚠️  Missing: ${missing}`);
  console.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);
  console.log(`Skills ${dryRun ? 'would be' : 'are now'} in: ${target}`);
  if (!dryRun && fs.existsSync(path.join(target, 'code-review'))) {
    if (withCli) fetchReviewCli(target);
    else console.log(`ℹ️  code-review's review-cli binary is not bundled — re-run with --with-cli to fetch it (needs gh auth; the repo is private). See README "Native binaries".`);
  }
}

main();
